src/Microsoft.AspNet.Mvc.Xml/RequiredValidationHelper.cs

```diff
@@ -0,0 +1,209 @@
+    /// </summary>
+    public class DataAnnotationRequiredAttributeValidation
+    {
+        private ConcurrentDictionary<Type, Dictionary<Type, List<string>>> _cachedValidationErrors
+            = new ConcurrentDictionary<Type, Dictionary<Type, List<string>>>();
+
+        public void Validate([NotNull] Type modelType, [NotNull] ModelStateDictionary modelStateDictionary)
+        {
+            var visitedTypes = new HashSet<Type>();
+
+            // Every node maintains a dictionary of Type => Errors. 
+            // It's a dictionary as we want to avoid adding duplicate error messages.
+            // Example:
+            // In the following case, from the perspective of type 'Store', we should not see duplicate
+            // errors related to type 'Address'
+            // public class Store
+            // {
+            //    [Required]
+            //    public int Id { get; set; }
+            //    public Address Address { get; set; }
+            // }
+            // public class Employee
+            // {
+            //    [Required]
+            //    public int Id { get; set; }
+            //    public Address Address { get; set; }
+            // }
+            // public class Address
+            // {
+            //    [Required]
+            //    public string Line1 { get; set; }
+            //    [Required]
+            //    public int Zipcode { get; set; }
+            //    [Required]
+            //    public string State { get; set; }
+            // }
+            var rootNodeValidationErrors = new Dictionary<Type, List<string>>();
+
+            Validate(modelType, visitedTypes, rootNodeValidationErrors);
+
+            foreach (var validationError in rootNodeValidationErrors)
+            {
+                foreach (var validationErrorMessage in validationError.Value)
+                {
+                    // Add error message to model state as exception to avoid
+                    // disclosing details to end user as SerializableError sanitizes the
+                    // model state errors having exceptions with a generic message when sending
+                    // it to the client.
+                    modelStateDictionary.TryAddModelError(
+                        validationError.Key.FullName,
+                        new InvalidOperationException(validationErrorMessage));
+                }
+            }
+        }
+
+        private void Validate(
+            Type modelType,
+            HashSet<Type> visitedTypes,
+            Dictionary<Type, List<string>> errors)
+        {
+            // Need not handle Key-Value pair specially here (for example, when the model type is Dictionary<,>
+            // which implements IEnumerable<KeyValuePair<TKey, TValue>>) as the model type here would be 
+            // KeyValuePair<TKey, TValue> where TKey and TValue are public properties which would also 
+            // be probed for Required attribute validation.
```

Comment:
probably my comment was not clear...I wanted to say that Key and Value are properties on KeyValuePairt<> type which would be probed,which in turn would cause probing of the types represented by these properties(i.e TKey and TValue respectively).